lib/src/rules/unnecessary_statement.dart

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.unnecessary_statement;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc =
+    r"Don't write unnecessary statements neither use getters with side-effect.";
+
+const _details = r'''
+
+**DON'T** use unnecessary statements neither use getters with side-effect.
+
+**BAD:**
+```
+void main() {
+  int a;
+  a; // LINT
+}
+```
+
+**GOOD:**
+```
+void main() {
+  int a;
+  foo(a);
+}
+```
+
+''';
+
+class UnnecessaryStatement extends LintRule {
+  _Visitor _visitor;
+  UnnecessaryStatement()
+      : super(
+            name: 'unnecessary_statement',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _UnnecessaryStatementVisitor extends UnifyingAstVisitor {
+  var isUnnecessaryStatement = true;
```

Comment:
I'd like to see this class renamed to be a bit more abstract. What this visitor is really doing, I think, is trying to determine whether we can prove that a given expression cannot have any side-effect. The fact that side-effect free expressions are "unnecessary" when they appear in certain locations is specific to this particular use case, but I'm sure you can imagine other uses cases for determining whether code has a side-effect.

You could leave the field with the current semantics and call it something like `hasNoSideEffect`, but for long term maintenance it would be better to remove the field and allow each `visit` method to return a `bool` indicating whether the visited node is side-effect free, with a fallback for unsupported expressions that returns `false` because that's safe.